### PR TITLE
[ refactor ] Experimenting with variables in `Data.List`

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
An experiment with the new `variable` key word informed by the discussion in #697. This was enormously satisfying to do :smile: Comments welcome.

### CODE REVIEW

File: src/Data/List/Base.agda [hunk 1]
```diff
@@ -217,7 +212,7 @@ allFin n = tabulate id
 
 -- Other
 
-unfold : ∀ {a b} {A : Set a} (B : ℕ → Set b)
+unfold : ∀ {A : Set a} (B : ℕ → Set b)
          (f : ∀ {n} → B (suc n) → Maybe (A × B n)) →
```
Comment (gallais) [inline src/Data/List/Base.agda]: Could this be: `unfold : (P :  ℕ → Set p) (...)` instead?

File: src/Data/List/Base.agda [hunk 2]
```diff
@@ -228,130 +223,117 @@ unfold B f {n = suc n} s with f s
...
+splitAt : ℕ → List A → (List A × List A)
 splitAt zero    xs       = ([] , xs)
 splitAt (suc n) []       = ([] , [])
 splitAt (suc n) (x ∷ xs) with splitAt n xs
 ... | (ys , zs) = (x ∷ ys , zs)
 
-takeWhile : ∀ {a p} {A : Set a} {P : Pred A p} →
-            Decidable P → List A → List A
+takeWhile : ∀ {p} {P : Pred A p} → Decidable P → List A → List A
 takeWhile P? []       = []
```
Comment (gallais) [inline src/Data/List/Base.agda]: You can drop `{p}` for all of these (or should remove `p` from the list of Level variables).

**Comment by gallais** [comment]
Beautiful!

File: src/Data/List/Base.agda [hunk 3]
```diff
@@ -144,65 +144,60 @@ sum = foldr _+_ 0
...
-  applyDownFrom f zero = []
-  applyDownFrom f (suc n) = f n ∷ applyDownFrom f n
+applyDownFrom : (ℕ → A) → ℕ → List A
+applyDownFrom f zero = []
+applyDownFrom f (suc n) = f n ∷ applyDownFrom f n
 
-  tabulate : ∀ {n} (f : Fin n → A) → List A
-  tabulate {zero}  f = []
-  tabulate {suc n} f = f Fin.zero ∷ tabulate (f ∘ Fin.suc)
+tabulate : ∀ {n} (f : Fin n → A) → List A
```
Comment (m0davis) [inline src/Data/List/Base.agda]: Couldn't you drop `∀ {n}` and it means the same thing? I believe that then you can also drop `n = ` from the lhs below.

**Comment by JacquesCarette** [comment]
That does represent a serious improvement in readability. Very nice.

File: src/Data/List/Base.agda (hunk 3 above)
Comment (MatthewDaggitt) [reply src/Data/List/Base.agda]: Yup we could. However the conclusion of #697 seems to be that until we have a mechanism in emacs for displaying a term's full type at the site of its definition, we're going to stick with only generalising things of type `Level` and `Set` in order to avoid the user having hold to much in their head at once. This policy will probably be relaxed later to include things like naturals.

**Comment by MatthewDaggitt** [comment]
Okay, I've done the same for `Data.Nat.Properties`. The only place I'm not sure about is the algebraic structures for `_++_` where I've kept the anonymous module with the `Set` parameter in order to better open the algebraic structures package. I'm not sure which people think is clearer:
```agda
module _ {A : Set a} where
  open FunctionProperties {A = List A} _≡_
  open Structures         {A = List A} _≡_

  ++-assoc : Associative _++_
  ...

  ++-identityˡ : LeftIdentity [] _++_
  ...
```
or
```agda
++-assoc : Associative {A = List A} _≡_ _++_
...

++-identityˡ : LeftIdentity {A = List A} _≡_ [] _++_
...
```
Otherwise if everyone is happy with this, I'll update the style guide and people are free to start adding variables to existing files.

**Comment by JacquesCarette** [comment]
In this particular case shown above, I think I prefer the second (i.e. without the anonymous module)! But see below where I give a way to 'count' to reverse that position.

Though, in general, I tend to prefer using a Minimum Description Length based heuristic: factor out commonalities when there is an overall savings.  Roughly speaking, the anonymous module adds about 25 tokens, and to save 6 tokens per line.  So as soon as you use > 4 things, you're starting to save. 

In other words, for me, it depends on how much actual repetition there is.  Diving into this case i.e. `Data.List.Properties`, you've got 6 (named) properties and a few structures, enough to actually make the anonymous module pay for itself!

**Comment by MatthewDaggitt** [comment]
I'll push a few commits updating the style guide and the most common modules. Otherwise people are free to start using them in pull requests!
